docs/library/machine.CAN.rst

```diff
@@ -0,0 +1,364 @@
+
+The ``machine.CAN`` interface is a *low level basic* CAN messaging interface
+that abstracts a CAN controller as an outgoing priority queue for sending
+messages, an incoming queue for receiving messages, and mechanisms for reporting
+errors.
+
+.. note:: The forthcoming ``can`` and ``aiocan`` micropython-lib modules are the
+          recommended way to use CAN with MicroPython.
+
+Constructor
+-----------
+
+.. class:: CAN(id, **kwargs)
+
+   Construct a CAN controller object of the given id:
+
+    - ``id`` identifies a particular CAN controller object; it is board and port specific.
+    - ``**kwargs`` are hardware-specific identifiers for the particular
+      board and port. For example, to set which GPIO pins to use for the
+      CAN controller.
+
+Methods
+-------
+
+.. method:: CAN.init(bitrate, mode=CAN.Mode.NORMAL, sample_point=75, sjw=None, tseg1=None, tseg2=None, f_clock=None)
+
+   Initialise the CAN bus with the given parameters:
+
+   - ``bitrate`` is the desired bus bit rate in bits per second.
+   - ``mode`` is one of :class:`CAN.Mode` enumerated values, indicating the
+     desired mode of operation.
+   - ``sample_point`` is a percentage of the data bit time. It specifies the
+     position of the bit sample with respect to the whole nominal bit time.
+
+   The remaining parameters all relate to CAN bit timings. If not specified, the CAN controller
+   will pick nominal values based on the specified bit rate.
+
+   Otherwise, they can be specified:
+
+    - ``sjw`` is the resynchronisation jump width in units of time quanta for nominal bits;
+      it can be a value between 1 and 4 inclusive for classic CAN, and between 1 and 128 inclusive for CAN FD.
+    - ``tseg1`` defines the location of the sample point in units of the time quanta for nominal bits;
+      it can be a value between 1 and 16 inclusive for classic CAN, and between 2 and 256 inclusive for CAN FD.
+    - ``tseg2`` defines the location of the transmit point in units of the time quanta for nominal bits;
+      it can be a value between 1 and 8 inclusive for classic CAN, and between 2 and 128 inclusive for CAN FD.
+    - ``f_clock`` is the system clock frequency for the CAN peripheral. This
+      can be omitted on many ports where the value is known to the hardware.
+
+    In this case, the CAN controller is configured correctly for the desired
+    ``bitrate`` and the specified total number of time quanta per bit.
+
+.. method:: CAN.init_fd(bitrate, sample_point=None, sjw=None, tseg1=None, tseg2=None, f_clock=None)
+
+   Initialise CAN FD for controllers which support it.
+
+   .. note:: This method is not present on the class for controllers that only
+             support CAN Classic.
+
+   This must be called after :func:`CAN.init()`. The parameters all have the
+   same meanings and behaviour as that function, however they configure the
+   Bit Rate Switch (BRS) feature of CAN FD.
+
+   When specifying exact timing parameters, the accepted ranges are:
```

Comment:
Same as comment above. How this works makes sense. On the imxrt I'd have to deinit and then reinit the can though after this was called. But, I could keep the regular settings. As mentioned above, giving tseg1/tseg2 control is something few people will use and it's easier to use the abstracted sample_point api as that's what you are trying to achieve by adjusting the segment times.

Reply:
> giving tseg1/tseg2 control is something few people will use

We can probably roll this into the above comments, with the goal that whatever API is available in `init()` for CAN Classic will be mirrored in `init_fd()` for CAN FD timings.

> On the imxrt I'd have to deinit and then reinit the can though after this was called. But, I could keep the regular settings.

I was just reading over this part in the iMX datasheet. I agree this is a bit fiddly in this case, sorry about that! On balance, it still seems better than trying to jam all the parameters into one place though.

Your comment did also get me thinking through the implications of having a CAN FD controller come online for a short period in an incompatible mode, before it reconfigures to support `BRS`. I think the way around this will be for the high-level `can` Python class to call `init(mode=CAN.Mode.SLEEP, ...)` and then `init_fd(mode=CAN.Mode.NORMAL, ...)` so the "Classic CAN" initialisation doesn't result in a transient period where the controller is active on the bus.